### PR TITLE
Nginx 1.15.0 & 1.16.1

### PR DESCRIPTION
**Pull request by julek-wolfssl**
This PR updates support for nginx 1.15.0 and adds support for 1.16.1. Some of the more important changes are:
- in wolfSSL_OCSP_resp_find_status set thisupd and nextupd to correct WOLFSSL_ASN1_TIME values
- src/ssl.c
-- check if we are overwriting an existing session id. If we are then don't overwrites certificates unless present in new version of the session.
-- implement parsing of ASN date format for thisDate and nextDate

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -12286,7 +12288,27 @@ int AddSession(WOLFSSL* ssl)
...
-        idx = SessionCache[row].nextIdx++;
+        for (i=0; i<SESSIONS_PER_ROW; i++) {
+            if (ssl->options.tls1_3) {
+            	if (XMEMCMP(ssl->session.sessionID, SessionCache[row].Sessions[i].sessionID, ID_LEN) == 0) {
+            		WOLFSSL_MSG("Session already exists. Overwriting.");
+            		overwrite = 1;
+            		idx = i;
+            		break;
+            	}
+            } else {
```
Comment (SparkiDev) [inline src/ssl.c]: formating

File: src/ssl.c [hunk 2]
```diff
@@ -12286,7 +12288,27 @@ int AddSession(WOLFSSL* ssl)
             return BAD_MUTEX_E;
         }
 
-        idx = SessionCache[row].nextIdx++;
+        for (i=0; i<SESSIONS_PER_ROW; i++) {
+            if (ssl->options.tls1_3) {
+            	if (XMEMCMP(ssl->session.sessionID, SessionCache[row].Sessions[i].sessionID, ID_LEN) == 0) {
```
Comment (SparkiDev) [inline src/ssl.c]: Don't use tab.

File: src/ssl.c [hunk 3]
```diff
@@ -12286,7 +12288,27 @@ int AddSession(WOLFSSL* ssl)
             return BAD_MUTEX_E;
         }
 
-        idx = SessionCache[row].nextIdx++;
+        for (i=0; i<SESSIONS_PER_ROW; i++) {
+            if (ssl->options.tls1_3) {
```
Comment (SparkiDev) [inline src/ssl.c]: Same code regardless of tls1_3???

File: src/ssl.c [hunk 4]
```diff
@@ -12360,9 +12382,15 @@ int AddSession(WOLFSSL* ssl)
 
 #ifdef SESSION_CERTS
     if (error == 0) {
-        session->chain.count = ssl->session.chain.count;
-        XMEMCPY(session->chain.certs, ssl->session.chain.certs,
-                sizeof(x509_buffer) * MAX_CHAIN_DEPTH);
+    	if (!overwrite || (overwrite && ssl->session.chain.count > 0)) {
```
Comment (SparkiDev) [inline src/ssl.c]: Why overwrite when !overwrite?

File: src/ssl.c [hunk 5]
```diff
@@ -37565,17 +37652,77 @@ long wolfSSL_ctrl(WOLFSSL* ssl, int cmd, long opt, void* pt)
...
+                /* On successful setting of new chain up all refs */
+                if (wolfSSL_X509_up_ref(x509) != 1) {
+                    WOLFSSL_MSG("Error increasing reference count");
+                    continue;
+                }
+            }
+        }
+    }
+#else
+    WOLFSSL_MSG("Session certificates not compiled in");
```
Comment (SparkiDev) [inline src/ssl.c]: Indent 4 more spaces

File: src/ssl.c [hunk 6]
```diff
@@ -38678,17 +38828,19 @@ int wolfSSL_X509_NAME_print_ex(WOLFSSL_BIO* bio, WOLFSSL_X509_NAME* name,
             }
             totalSz += tmpSz;
         }
+        if (totalSz > 0 && fullName[totalSz-1] == '\0')
```
Comment (SparkiDev) [inline src/ssl.c]: Why is this necessary?
Is this something that can be fixed above?

**Review by SparkiDev — CHANGES_REQUESTED** [review]
Make sure you fix up the formatting and remove TABs.

File: src/ssl.c (hunk 3 above)
Comment (julek-wolfssl) [reply src/ssl.c]: Not the same. For TLS 1.3 the session is stored in `ssl->session` while for other protocols `ssl->arrays` is used.

File: src/ssl.c (hunk 4 above)
Comment (julek-wolfssl) [reply src/ssl.c]: If we are not overwriting then we are writing to a new session object (or the next one to be used according to `idx = SessionCache[row].nextIdx++;`). Overwriting means that we have found a session with the same `sessionID` and we want to update the data in the old session.

File: src/ssl.c (hunk 6 above)
Comment (julek-wolfssl) [reply src/ssl.c]: As per my comment above the function:
> The BIO output of  wolfSSL_X509_NAME_print_ex does NOT include the null terminator
The null terminator is automatically appended by both `XSNPRINTF` and `XSTRNCAT`. Both also require C strings as input so removing the null terminator at that stage is not possible. The simplest solution I found is decrementing the `totalSz` at the end.

Comment (julek-wolfssl) [reply src/ssl.c]: Actually nevermind. I found a way.

**Comment by julek-wolfssl** [comment]
I updated my PR according to your requests. @SparkiDev please take a look at my replies.

**Review by SparkiDev — APPROVED** [review]

**Review by toddouska — APPROVED** [review]

**Comment by toddouska** [comment]
Looks good, please resolve merge conflicts though.

**Review by toddouska — APPROVED** [review]

**Pull request merged**